### PR TITLE
Ownership docs

### LINKED ISSUE #9: Document `collections` API
Namely, we should be explicit about "ownership" re: our data structures.

### LINKED ISSUE #11: Refactor JSON object creation duplicated code
(Kevan knows what this means. lol)

`generate_ACK_msg`

### PR DESCRIPTION
**Pull request by khollbach**
Getting fancy with github (maybe I'll regret it, lol) -- these changes are based on the current state of the `shared/tcp` branch. Let me know what you think! I think we'd be able to merge these into that branch, and then later merge that branch into main.

### CODE REVIEW

File: src/challenge-2/challenge-2.c [hunk 1]
```diff
@@ -34,7 +31,7 @@ int main(void)
     const char** peers = node_ids(init_msg);
     const size_t num_peers = node_ids_count(init_msg);
```
Comment (Per48edjes) [inline src/challenge-2/challenge-2.c:32]: QQ: Why is `node_ids_count` returning a `size_t`, but here we make `num_peers` a `const size_t` (versus original approach of making `node_ids_count` return a `const size_t`)?

File: lib/collections.c [hunk 2]
```diff
@@ -30,13 +30,14 @@ typedef struct Queue
     Node* tail;
     size_t length;
     size_t max_length;
+    void (*free_function)(void*);
```
Comment (Per48edjes) [inline lib/collections.c:33]: Oh, very nice! 

Thinking out loud (for my understanding): the queue 'owns' its comprising data, so it should free it accordingly; the data is polymorphic, so the appropriate `free_function` varies depending on what's in the queue.

File: lib/tcp.h [hunk 3]
```diff
@@ -2,9 +2,14 @@
 
 #include <json-c/json.h>
 
+// Takes ownership of `msg`.
```
Comment (Per48edjes) [inline lib/tcp.h:5]: For my edification: This is 'ownership' insofar as the sibling function (`msg_recv_listener`) ultimately frees it [here](https://github.com/khollbach/gossip-glomers-c/blob/d036b168066affb9396889e3fae787da2e64062c/lib/tcp.c#L117)?

File: lib/util.c [hunk 4]
```diff
@@ -73,9 +73,7 @@ json_object* generic_reply(json_object* msg)
     json_object_object_add(reply_body, "in_reply_to", reply_in_reply_to);
 
     // Generate reply type: "XXX_ok"
-    json_object* type =
-        json_object_object_get(json_object_object_get(msg, "body"), "type");
-    const char* name = json_object_get_string(type);
+    const char* name = msg_type(msg);
```
Comment (Per48edjes) [inline lib/util.c:76]: 😆  I didn't even notice you'd implemented this convenience function already!

File: src/challenge-2/challenge-2.c [hunk 5]
```diff
@@ -76,7 +74,7 @@ void follower_event_loop()
 
 void leader_event_loop()
 {
-    Queue* queue = queue_init(5);
+    Queue* queue = queue_init(5, queue_json_object_free);
```
Comment (Per48edjes) [inline src/challenge-2/challenge-2.c]: Optional: Might be worth renaming this `conch_request_queue` just to disambiguate what queue this is (given we've got queues all over the place now).

File: src/challenge-2/challenge-2.c [hunk 6]
```diff
@@ -91,15 +89,18 @@ void leader_event_loop()
...
-                assert(!conch_is_available(conch));
                 fprintf(stderr, "Warn: leader_event_loop: shut down with "
-                                "unserved conch requests\n");
+                                "unreturned conch\n");
             }
-            break;
+
+            // Shutdown.
+            queue_free(queue);
+            conch_free(conch);
```
Comment (Per48edjes) [inline src/challenge-2/challenge-2.c]: Good catch!

**Review by Per48edjes — APPROVED** [review]
Awesome work, @khollbach ! The annotations will be super helpful (e.g., using `clangd` LSP surfaces these when asking for type info).

Feel free to merge this into `shared/tcp` (or rebase `shared/tcp` onto this)!

File: lib/collections.c (hunk 2 above)
Comment (khollbach) [reply lib/collections.c:33]: :100:

File: lib/tcp.h (hunk 3 above)
Comment (khollbach) [reply lib/tcp.h:5]: Exactly! Locally-speaking, ownership first goes into the function `msg_send_pusher` (specifically the local variable `msg` "owns" the data). And then inside that function, ownership transfers to the `Message` object as the `->msg` field. And then into the `Queue` when we call `enqueue`.

When we pop that particular element from the queue in `recv_listener`, ownership is passed into that function, which immediately frees it. (I'd even say that `free` takes ownership, just permanantly.)

File: lib/util.c (hunk 4 above)
Comment (khollbach) [reply lib/util.c:76]: np! I think we might've introduced the convenience function after this code was written :)

File: src/challenge-2/challenge-2.c (hunk 5 above)
Comment (khollbach) [reply src/challenge-2/challenge-2.c]: Done!

**Pull request merged**